i18n-helpers/src/lib.rs

```diff
@@ -54,33 +56,215 @@ pub fn extract_paragraphs(text: &str) -> impl Iterator<Item = (usize, &str)> {
+
+    /// True if we are within a codeblock right now.
+    in_codeblock: bool,
+
+    /// True if the last message can still be appended to. If this is true then
+    /// msgs has at least one element.
+    message_open: bool,
+}
+
+impl<'a> MsgAccumulator<'a> {
+    fn new(text: &'a str) -> Self {
+        Self {
+            text: text,
+            offsets: text.match_indices("\n").map(|(offset, _)| offset).collect(),
+            msgs: vec![],
+            in_codeblock: false,
+            message_open: false,
+        }
+    }
+
+    /// Mark the current message as finished.
+    fn finish_message(&mut self) {
+        println!("finish");
+        self.message_open = false;
+    }
+
+    /// Add a new text message, or extend an existing one.
+    fn push_message(&mut self, msg: Message) {
+        println!("push_message({msg:?})");
+        if self.message_open {
+            if let Some(last) = self.msgs.last_mut() {
+                if last.try_push(&msg) {
+                    return;
+                }
+            }
+        }
+        if matches!(msg, Message::Any(_)) {
+            panic!("Tried to push an Any value with no previous message");
+        }
+        self.msgs.push(msg);
+        self.message_open = true;
+    }
+
+    /// Calculate the line number for the given offset.
+    fn line_number(&self, offset: usize) -> usize {
+        self.offsets.partition_point(|&o| o < offset) + 1
+    }
+
+    /// Push a new Markdown event into the accumulator.
+    fn push_event(&mut self, evt: Event, offset: Range<usize>) {
+        dbg!(&offset);
+        match dbg!(evt) {
+            Event::Start(Tag::Emphasis | Tag::Strong | Tag::Strikethrough) => {
+                self.push_message(Message::Text(offset))
+            }
+            Event::Start(Tag::CodeBlock(_)) => {
+                // note that this offset contains the entire codeblock!
+                self.push_message(Message::Code(offset));
+                self.in_codeblock = true;
+            }
+            Event::Start(_) => self.finish_message(),
+
+            Event::End(Tag::Emphasis | Tag::Strong | Tag::Strikethrough) => {
+                self.push_message(Message::Text(offset))
+            }
+            Event::End(Tag::CodeBlock(_)) => {
+                self.in_codeblock = false; // (assumes codeblocks aren't nested!)
+                self.finish_message();
+            }
+            Event::End(_) => self.finish_message(),
+
+            Event::Text(_) if !self.in_codeblock => {
+                self.push_message(Message::Text(offset))
+            }
+            Event::Code(content) => self.push_message(Message::Code(offset)),
+            Event::Html(content) => self.push_message(Message::Html(offset)),
+            // TODO: list items
+            // TODO: headings
+            _ => {}
+        }
+    }
+
+    /// Get the resulting list of messages.
+    fn into_msgs(mut self) -> Vec<(usize, &'a str)> {
+        let parser = Parser::new(self.text);
```

Comment:
You will eventually want to use [`new_cmark_parser`](https://docs.rs/mdbook/latest/mdbook/utils/fn.new_cmark_parser.html) here to have the same Markdown extensions enabled as in mdbook.